### PR TITLE
Scripting: Convert script fields to use script context

### PR DESCRIPTION
**Pull request by rjernst**
This commit removes the use of SearchScript for script fields and adds
a new FieldScript.

### CODE REVIEW

File: docs/painless/painless-contexts/painless-field-context.asciidoc [hunk 1]
```diff
@@ -14,18 +14,15 @@ a customized value for each document in the results of a query.
         Contains the fields of the specified document where each field is a
         `List` of values.
 
-{ref}/mapping-source-field.html[`ctx['_source']`] (`Map`)::
+{ref}/mapping-source-field.html[`params['_source']`] (`Map`)::
```
Comment (jdconrad) [inline docs/painless/painless-contexts/painless-field-context.asciidoc]: Do you mind adding read-only to this, like how _score was?  This shouldn't be modified as part of a script field.  My mistake from the first iteration of these docs.

File: server/src/main/java/org/elasticsearch/script/FieldScript.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ */
+public abstract class FieldScript {
+
+    private static final Map<String, String> DEPRECATIONS;
+
+    static {
+        Map<String, String> deprecations = new HashMap<>();
+        deprecations.put(
+            "doc",
+            "Accessing variable [doc] via [params.doc] from within a terms-set-query-script " +
```
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/script/FieldScript.java]: Think this was a copy and paste issue -- should be field-script and not terms-set-query

**Review by jdconrad — APPROVED** [review]
LGTM.  Thanks for doing this!

**Comment by rjernst** [comment]
@jdconrad I addressed your comments. I will wait for https://github.com/elastic/elasticsearch/pull/33856 to get in so I can re-use the expression script engine work there to add expression support of the new field script.

**Pull request merged**